src/pages/Messages/_Star.js

```diff
@@ -0,0 +1,38 @@
+import React from 'react'
+import { connect } from 'react-redux'
+import Tooltip from '../../components/Tooltip'
+import { Button } from 'antd'
+import {deleteStar, unstarMessage, starMessage, submitStar} from "../../actions/starMessage"
+
+const Star = ({currentUser, isStarred, toggleStar}) => {
+  const handleStarClick = (event) => {
+    if(event){ event.preventDefault() }
+    toggleStar()
+  }
+  return (
+
+    <Tooltip placement='top' title="Star this message">
+      <Button
+        icon={isStarred ? 'star' : 'star-o'}
+        style={isStarred ? {color: '#ffd700'} : {color: '#000000'}}
+        onClick={handleStarClick}
+        />
+    </Tooltip>
+
+  )
+}
+
+const mapStateToProps = () => ({
+})
+
+const mapDispatchToProps = (dispatch, {isStarred, messageId}) => ({
+  toggleStar: () => (
+      isStarred ? dispatch(unstarMessage(messageId)) & dispatch(deleteStar(messageId)) : dispatch(starMessage(messageId)) & dispatch(submitStar(messageId))
```

Comment:
Looks like we're missing a few ampersands, e.g. `&` should be `&&`.

Which is a good example of how ternary operators are hard to read when they get really long. Doubly so when they start doing multiple things per case.

I'd either use a variable to wrap the actions so it's a single action in the ternary:

```js
toggleStar: () => {
  const starMessage = () => {
    dispatch(unstarMessage(messageId))
    dispatch(deleteStar(messageId))
  }
  const unstarMessage = () => {
    dispatch(starMessage(messageId))
    dispatch(submitStar(messageId)
  }

  return isStarred ? unstarMessage() : starMessage()
}
```

Or just use a plain `if / else`:


```js
toggleStar: () => {
  if (isStarred) {
    dispatch(unstarMessage(messageId))
    dispatch(deleteStar(messageId))
  } else {
    dispatch(starMessage(messageId))
    dispatch(submitStar(messageId)
  }
}
```

Reply:
Whichever format you chose, the behavior needs to be tweaked too. We don't want to change the client state until we know the server action has succeeded.

To do that, we can rearrange this to use callbacks:

```js
toggleStar: () => {
  const starMessage = () => {
    const onSuccess () => dispatch(unstarMessage(messageId))

    return dispatch(deleteStar(messageId, onSuccess))
  }
  const unstarMessage = () => {
    const onSuccess = () => dispatch(starMessage(messageId))

    return dispatch(submitStar(messageId, onSuccess))
  }

  return isStarred ? unstarMessage() : starMessage()
}
```

Reply:
I tried to implement this suggestion but for some reason, the way you suggested does not ever hit the starredMessage actions I put a console.log within the reducer and only saw it hit when my toggleStar method was ` toggleStar: () => (
     isStarred ? dispatch(unstarMessage(messageId)) && dispatch(deleteStar(messageId)) : dispatch(starMessage(messageId)) && dispatch(submitStar(messageId))
 )` , it does submit the star successfully to the server side though

Reply:
Ok, scratch that above comment,  I had a problem on the server side. I didnt properly return `:ok` and `:error` replies from my create and delete handlers, so the onSuccess conditional wasn't being hit on the client side here.